src/Compilers/CSharp/Test/Semantic/Semantics/StaticNullChecking.cs

```diff
@@ -32342,9 +32284,312 @@ static void F1<T1>(T1 t1)
+class C<T, U>
+    where T : U?
+    where U : T?
+{
+}
+class D<T, U>
+    where T : class, U?
+    where U : class, T?
+{
+}
+class E<T, U>
+    where T : class, U
+    where U : T?
+{
+}";
+            var comp = CreateCompilation(source, parseOptions: TestOptions.Regular8);
+            comp.VerifyDiagnostics(
+                // (10,9): error CS0454: Circular constraint dependency involving 'T' and 'U'
+                // class C<T, U>
+                Diagnostic(ErrorCode.ERR_CircularConstraint, "T").WithArguments("T", "U").WithLocation(10, 9),
+                // (15,9): error CS0454: Circular constraint dependency involving 'T' and 'U'
+                // class D<T, U>
+                Diagnostic(ErrorCode.ERR_CircularConstraint, "T").WithArguments("T", "U").WithLocation(15, 9),
+                // (20,9): error CS0454: Circular constraint dependency involving 'T' and 'U'
+                // class E<T, U>
+                Diagnostic(ErrorCode.ERR_CircularConstraint, "T").WithArguments("T", "U").WithLocation(20, 9));
+        }
+
+        [WorkItem(27289, "https://github.com/dotnet/roslyn/issues/27289")]
+        [Fact]
+        public void NullableTInConstraint_03()
+        {
+            var source =
+@"class A<T> where T : T, T? { }
+class B<U> where U : U?, U { }
+class C<V> where V : V?, V? { }";
+            var comp = CreateCompilation(source, parseOptions: TestOptions.Regular8);
+            comp.VerifyDiagnostics(
+                // (3,26): error CS0405: Duplicate constraint 'V' for type parameter 'V'
+                // class C<V> where V : V?, V? { }
+                Diagnostic(ErrorCode.ERR_DuplicateBound, "V?").WithArguments("V", "V").WithLocation(3, 26),
+                // (3,9): error CS0454: Circular constraint dependency involving 'V' and 'V'
+                // class C<V> where V : V?, V? { }
+                Diagnostic(ErrorCode.ERR_CircularConstraint, "V").WithArguments("V", "V").WithLocation(3, 9),
+                // (1,9): error CS0454: Circular constraint dependency involving 'T' and 'T'
+                // class A<T> where T : T, T? { }
+                Diagnostic(ErrorCode.ERR_CircularConstraint, "T").WithArguments("T", "T").WithLocation(1, 9),
+                // (1,9): error CS0454: Circular constraint dependency involving 'T' and 'T'
+                // class A<T> where T : T, T? { }
+                Diagnostic(ErrorCode.ERR_CircularConstraint, "T").WithArguments("T", "T").WithLocation(1, 9),
+                // (2,9): error CS0454: Circular constraint dependency involving 'U' and 'U'
+                // class B<U> where U : U?, U { }
+                Diagnostic(ErrorCode.ERR_CircularConstraint, "U").WithArguments("U", "U").WithLocation(2, 9),
+                // (2,9): error CS0454: Circular constraint dependency involving 'U' and 'U'
+                // class B<U> where U : U?, U { }
+                Diagnostic(ErrorCode.ERR_CircularConstraint, "U").WithArguments("U", "U").WithLocation(2, 9));
+        }
```

Comment:
>} [](start = 8, length = 1)

I don't remember the LDM decision on whether nullability constraints need to agree. For instance, `where T : I?, U`.

Reply:
We're not checking nullability of constraints currently, but according to the notes from 4/25, nullability of constraints should agree. `nullable-reference-types.md` contains a reference to those notes for now.

---
In reply to: [197943706](https://github.com/dotnet/roslyn/pull/27525#discussion_r197943706) [](ancestors = 197943706)